### ISSUE TITLE
How to prepare my own training datasets

### ISSUE THREAD

**Issue by kaihuchen** [tensorflow/tfjs-core#20]
I am trying to prepare my own datasets for use with the  model builder example, but it is not clear to me what is the right way to do so.

For example, in the config file deeplearnjs/demos/model-builder/model-builder-datasets-config.json there are specifications for "images" and "labels" for CIFAR-10, which point to https://storage.googleapis.com/learnjs-data/model-builder/cifar10_images.png and 
https://storage.googleapis.com/learnjs-data/model-builder/cifar10_labels_uint8 respectively. However, the images file does not appear to be a simple mosaic of the training images, and the labels file is not a simple text file containing plain-text labels. 

Any information about this is much appreciated!

**Comment by nsthorat** [comment]
Hi Kaihu,

The collage for the model builder is a flattened version, you can generate that collage with this script, you will need to write your uint8 ndarray to a file (numpy has a ndarray.tofile method which you can use):
https://github.com/PAIR-code/deeplearnjs/blob/master/scripts/convert_uint8_tensor_to_png.py

The uint8 labels file is simply the uint8 byte packed file using the ndarray.tofile on a uint8 ndarray. The type of the file is specified in the model-builder-datasets-config.json.

Hope this helps, let me know if you need anything else. Looking forward to seeing what you do with this, please do send a link when you have something interesting!

**Comment by nsthorat** [comment]
By the way, you can use all of these without the model-builder, if you like - all the building blocks that we're talking about above just by using this file:
https://github.com/PAIR-code/deeplearnjs/blob/master/demos/xhr-dataset.ts

Then you can just construct your own graph (model builder only allows sequential models, that is, a model of layers where the data gets completely transformed between operations).

**Comment by kaihuchen** [comment]
@nsthorat  Thanks for your prompt reply. 

FYI, what I am doing here is a preliminary study to see if I am able to use deeplearnjs to build a ML-based tool suitable for a large number of non-programmers to use (e.g., a bug or plant disease recognizer that a farming expert can train, and then to be used by farmers eventually) . Deeplearnjs is ideal for my purposes, since it will allow such a system to scale way better/cheaper than the typical server-centric ML setup. 

There is something I still do not quite understand about the 'collage' issue. Based on your description and the code (i.e., convert_uint8_tensor_to_png.py) it seems that the assumed input is an array of 3D tensors. However, what I have to start with is actually a set of JPEG/PNG files instead. I wonder if there is any tool that will allow me to convert such JPEG/PNG files to the target flattened version, without having to convert them to tensors first. Thanks!

**Comment by nsthorat** [comment]
Wow that is super cool!

Ah yes, we assume a numpy 3d array. You could modify that script to read in your images through [scipy](https://docs.scipy.org/doc/scipy-0.13.0/reference/generated/scipy.ndimage.imread.html) or [PIL](https://stackoverflow.com/questions/384759/pil-and-numpy), or you could write a script that uses those to aggregate all of the numpy arrays with the methods above and write the resulting aggregated numpy array to a file. Then, you could run our scripts directly without changes.

I would probably go with the first approach, but it's up to you.

Hope this helps, let me know if you have any more questions!

**Comment by kaihuchen** [comment]
@nsthorat 
Found out that the labels must be encoded in the one-hot format (I think), which wasn't obvious to me earlier. At this point my dataset kind of runs normally, except that for some reasons it seems to be picking up only one class of examples during training.

More specifically, I have a dataset with three classes (A, B, and C) and a total of 534 training examples (as shown in the Model Build screenshot below). The model used here is a copy of CIFAR-10's but with the output changed from 10 to 3. The 'INFERENCE' column shows only one class (i.e., the class C) of examples (expecting 3 there). I noticed that if I reverse the order when flattening the training images, then it will still train on only one class (i.e., the class A).  Any help is appreciated!

![2017-08-10 11_02_51](https://user-images.githubusercontent.com/13685803/29177915-07acc0a2-7dbe-11e7-98b3-1ae8d617687e.png)

**Comment by kaihuchen** [comment]
Found some more clues here. The problem seems to have something to do with the size of the training set. 

I reduced the size of the training set to just 3 example (one per class, just to debug), everything looks in order, but the code somehow picks only one example for training. Same thing with 6 examples, or 534 examples. Replicated the images in the training set to yield 3000 examples, then the problem went away and all's well. Got the dataset with 534 examples and replicated it 5 times to yield 2670 example, and it also worked well. 

Now that I know how to work around it, it is not really a problem, since I am also supposed to be using larger training dataset anyways. Hope this will help somebody.

**Comment by nsthorat** [comment]
Ah glad you got it working, I will investigate why smaller datasets don't seem to work.

I'll also spend a little time making it easier to plug into the model-builder, we never really spent the cycles to make it extensible :)

**Comment by nsthorat** [comment]
@kaihuchen to expedite this, could you send me the dataset with only a few examples (the one that breaks the model builder)?

Thanks

**Comment by kaihuchen** [comment]
Attached are the data files (labels and images) for 3 classes of flowers (named simply as 'A', 'B', and 'C'), with two samples per class for a total of 6 samples. Hope this helps. 

![images](https://user-images.githubusercontent.com/13685803/29293878-c195f858-811a-11e7-92ab-7273ef05db98.png)

**Comment by kaihuchen** [comment]
@nsthorat Just realized that my *labels* files might not have gone through when I attached it, since I can't see it in my post. Wonder if you did get it?

**Comment by JainVikas** [comment]
@nsthorat  @kaihuchen, for one hot encoding for labels, i noticed that for CIFAR10, n_value seems to be set at "60000", should i follow same for encoding my labels, i have only 5 though, or "auto" n_value will work same? I am not able to run and check due to issue#50. Any guidance would be helpful, thanks

**Comment by kaihuchen** [comment]
@JainVikas   Not sure what you meant by 'n_value', but for the one-hot encoding of the labels I usually set the size of each one-hot vector to be the same as the number of classes in the data. So if you have 5 classes then using an one-hot vector of size 5 should work. Overall in the 'labels' file there should be N (N=number of data samples) such vectors laid out sequentially in the file (for a total of 5*N bytes, assuming 5 classes), where each byte should be an uint8 of either 0 or 1. 

Hope this helps!

**Comment by nsthorat** [comment]
I didn't get the labels. Could you send it to me on a private github mail?

What @kaihuchen said is right. Regarding #50, we will be fixing that bug soon (once we get our hands on a windows machine).

**Comment by hbtech-ai** [comment]
How to make dataset use my own images

**Comment by digitalheir** [comment]
I'm also trying to make my own dataset. I don't have experience with Python nor TensorFlow. So not exactly sure what to do. Does this approach make sense? 

## Generate inputs
```
    paths_to_images = [
        'a.png',
        'b.png',
        ... etc
    ]
    
    loaded_images = []    
    for path_to_image in paths_to_images:
        image = Image.open(path_to_image)
        image_as_array = np.asarray(image)  # width x height x channels
        loaded_images.append(image_as_array)

    a = np.array(loaded_images)

    print('Read', a.shape[0], 'images')
    print('min/max pixel values: ', a.min(), '/', a.max())

    # Make each image take a single row in the big batch image by flattening the
    # width (2nd) and height (3rd) dimension.
    # a has shape N x (Width*Height) x Channels.
    a = a.reshape([a.shape[0], -1, FLAGS.num_channels]).squeeze()
    im = Image.fromarray(a)
    out_path = 'output' + '.png'
    im.save(out_path)
    print('Saved image with width/height', im.size, 'at', out_path)
```

I guess this works correctly.

## Generate labels
* ??? Some way to encode an n-dimensional array in binary ???

## Import in model-builder

* Add data to `model-builder-datasets-config.json`
* ??? Some way to re-use model-builder-demo.html ???

I've tried to run `npm run prep` followed by `npm run build`. This puts compiled JavaScript in the `build` folder and downloads dependencies from bower. I'm unsure how to link this to `model-builder-demo.html`? If I open `model-builder-demo.html` directly, I get an error it can't find `deeplearnjs/demos/model-builder/bundle.js `.

**Comment by JainVikas** [comment]
@digitalheir @hbtech-ai it does, as a matter of fact, I have used the same approach.
**Generate inputs**
Read image files and create a combined ndarray for all images.
ndarray to be passed to create a flattened image, where ndrray will be a consolidated file for all images.
The collage for the model builder is a flattened version, you can generate that collage with this script, you will need to write your uint8 ndarray to a file (numpy has a #ndarray.tofile method which you can use): https://github.com/PAIR-code/deeplearnjs/blob/master/scripts/convert_uint8_tensor_to_png.py

**Generate labels**
use onehotencoder to generate labels file.
http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html

**Modify model builder demo**
add to model-builder-datasets-config.json
`"myTestData": {
    "data": [{
      "name": "images",
      "path": "<<flattened image file path>>"
      "dataType": "png", 
      "shape": [32, 32, 3]
    }, {
      "name": "labels",
      "path": << encoded labels files>>",
      "dataType": "uint8",
      "shape": [10]
    }],
    "labelClassNames": ["airplane", "automobile", "bird", "cat", "deer", "dog", "frog", "horse", "ship", "truck"],
    "modelConfigs": {
      "Convolutional": {
        "path": "cifar10-conv.json"
      }
    }
	}`

**Comment by hbtech-ai** [comment]
@digitalheir  I use you method to make dataset, but now is error like 
![_20170903132027](https://user-images.githubusercontent.com/27670432/30000797-ac4c6474-90aa-11e7-917d-89d0ee9857da.png)
@nsthorat I hope you can give developer some help for make dataset

**Comment by digitalheir** [comment]
@hbtech-ai I had an error in the way that I encoded the labels file - you need to specify to `numpy` that you work with `uint8` numbers. If you get this error, set a breakpoint in `getNDArray()` in `xhr-dataset.ts` to inspect whether the number of examples and number of labels is equal. I got it to work in the end, at least with some dummy data. Will check if I get to do real training later this week.

**Comment by hbtech-ai** [comment]
@digitalheir Thanks for your help,I find labels is not unit8

**Comment by hbtech-ai** [comment]
@digitalheir my own dataset is work,Thanks

**Comment by kaihuchen** [comment]
If it helps anybody, here is a little **Python program that I wrote for converting the training dataset to a form as required by deeplearnjs** :  http://www.smesh.net/docs/deeplearnjs/preprocess.py
The program is not polished, and there are certainly lots of room for further improvement. If anybody wants to take it and improve upon it, you are very much welcome to do so.

**Comment by nsthorat** [comment]
Sorry I'm late to this thread.

@digitalheir Are you interested in formalizing this into a tutorial that we host? I think that would be really valuable for others.

@kaihuchen Would you want to check that script into our scripts directory? I think others would find that useful. BTW, we're going to make better tooling around bringing data into the browser in the near future.

**Comment by digitalheir** [comment]
@nsthorat sure. But I think it's only really valuable once we can save trained weights. Is there any word on that yet? (https://github.com/PAIR-code/deeplearnjs/issues/33#issuecomment-323891195)

**Comment by nsthorat** [comment]
@digitalheir to be clear, you need weight saving for the model builder or for arbitrary graph code?

**Comment by digitalheir** [comment]
For the model builder. I think the model builder is a great tool to quickly prototype any machine learning project.

**Comment by kaihuchen** [comment]
@nsthorat  FYI, I have created a pull request for the said script. Let me know if any change is needed.

**Comment by digitalheir** [comment]
@nsthorat I'm ready to write a tutorial. I've prepared an example data set to recognize runes.

How do I write a tutorial? Add a markdown files to `gh-pages/docs/tutorials` & pull request?

**Comment by nsthorat** [comment]
@kaihuchen Awesome! I will review the PR tomorrow afternoon, we are preparing for a presentation in the morning now :)

@digitalheir Actually it's a little confusing right now - if you want to check in demo code, put it in a directory `master/demos/yourdemo` and put your markdown file in that directory. Our Jekyll setup is smart enough to crawl the demos and find the markdown files and index them. If you don't want to check in any code with it, put it in `master/docs/tutorials/yourtutorial` and then we'll run the make website script which will push the tutorial to the website (we're planning on moving away from gh-pages entirely pretty soon).

**Comment by kaihuchen** [comment]
@nsthorat    Good luck with your presentation! I hope that you will tell your audience that deeplearnjs is not just for quick prototyping, but rather it could lead to a paradigm change in large-scale distributed ML, as well as some serious business opportunities.

**Comment by digitalheir** [comment]
@nsthorat Please review

https://github.com/digitalheir/deeplearnjs/blob/rune_recognition_tutorial/demos/rune_recognition/rune_recognition.md

Should I also add the example images I've prepared to the repo? It's about 650kb. I thought it might bloat the code repository a bit too much.

**Comment by ZhimingHsu** [comment]
@kaihuchen  Thanks for your script
[process_images_for_training.py](https://github.com/PAIR-code/deeplearnjs/blob/master/scripts/process_images_for_training.py)
It help me a lot for my testing program.

I am working on OCR base on the rune recognition demo but input gray image
I found it need to add code for pack gray image.  

def preprocessImages(FLAGS):
=>  num_channels = FLAGS.num_channels //deal with gray image when set --num_channels=1 
=>  b = a.reshape([a.shape[0], -1, num_channels ]).squeeze()

**Comment by dexter1608** [comment]
hi how to create dataset as same format as mscoco?

**Comment by nsthorat** [comment]
TensorFlow.js has launched (https://js.tensorflow.org) and we will be working on data APIs in the coming months. Please re-open another issue in our new repo if you want to continue this discussion: https://github.com/tensorflow/tfjs
